[[ Bug 19307 ]] Invalidate children of deleted object

livecodeali opened the pull request:
When an object is deleted, all its children should be inaccessible
from the point of view of the object tree, so recursively invalidate
all the weak proxies as necessary.

Also ensure that players are closed and stopped when a parent is
deleted.

livecodeali committed:
[[ Bug 19307 ]] Invalidate children of deleted object
When an object is deleted, all its children should be inaccessible
from the point of view of the object tree, so recursively invalidate
all the weak proxies as necessary.

Also ensure that players are closed and stopped when a parent is
deleted.

livecodeali committed:
[[ MCplayers ]] Refactor MCplayers access
MCtemplateplayer belongs to MCplayers, and gets destructed and
re-constructed every time `reset the templateplayer` is called.

So the code that removes a player from MCplayers needs to happen
in the MCPlayer destructor. This means that when players have been
deleted and their proxies invalidated, it is possible for invalid
MCPlayerHandles to be part of the MCplayers list.

This patch refactors MCplayers access and, in particular, ensures
that no crash can occur when iterating through the MCplayers list.

montegoulding commented:
@livecodeali I think there's some stuff in `MCStack::del()` that should be moved to `MCStack::removereferences()`. 

```
if (MCtopstackptr == this)
    {
        MCtopstackptr = nil;
        MCstacks->top(nil);
    }
    if (MCstaticdefaultstackptr == this)
		MCstaticdefaultstackptr = MCtopstackptr;
	if (MCdefaultstackptr == this)
		MCdefaultstackptr = MCstaticdefaultstackptr;

	// MW-2008-10-28: [[ ParentScripts ]] If this stack has its 'has parentscripts'
	//   flag set, flush the parentscripts table.
	if (getextendedstate(ECS_HAS_PARENTSCRIPTS))
		MCParentScript::FlushStack(this);
    
    if (needs != NULL)
    {
        while (nneeds--)
            needs[nneeds]->removelink(this);
        
        delete needs;
        needs = NULL;
    }
  
    uint2 i = 0;
    while (i < MCnusing)
        if (MCusing[i] == this)
        {
            MCnusing--;
            uint2 j;
            for (j = i ; j < MCnusing ; j++)
                MCusing[j] = MCusing[j + 1];
        }
        else
            i++;
```

livecodeali committed:
[[ Obj Deletion ]] Move stack object tree removal related code

livecodeali committed:
[[ Bug 19447 ]] Delete temporary stacks using del
It was possible for temporary stack pointers to be deleted
directly while they or their children were outstanding message
targets. Calling `del` on them causes these messages to be
removed from the message queue.

`del` cannot be called on the current `home` stack (i.e the first
stack in `MCdispatcher`'s `stacks`), so we add a `dodel` function
to circumvent this when we really do want to delete this stack
internally.

peter-b commented:
@livecodeali I've had a look through this and it all seems sensible to me.

@livecode-vulcan review ok f6b11e9

livecode-vulcan commented:
💙 review by @peter-b ok f6b11e915975029f8e52085098ad02ca9e89df53

livecode-vulcan commented:
😎 test success f6b11e915975029f8e52085098ad02ca9e89df53

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/35)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/34)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/21)
* try-community-universal-ios-iphoneos10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/39)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/39)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/33)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/35)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/37)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/39)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/37)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/39)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/39)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/37)
